tools/migration-unified-annotation-versioning/migration.py

```diff
@@ -0,0 +1,529 @@
+            if tracing_id in mapping_id_map:
+                editable_mapping_id = mapping_id_map[tracing_id]
+                all_update_groups.append(self.fetch_updates(editable_mapping_id, "editableMapping", "editableMappingUpdates", json_encoder=json_encoder, json_decoder=json_decoder))
+
+        unified_version = 0
+        version_mapping = {}
+        for tracing_id, _ in layers:
+            version_mapping[tracing_id] = {0: 0} # We always want to keep the initial version 0 of all layers, even if there are no updates at all.
+
+        queue = []
+        for i, update_groups_for_layer in enumerate(all_update_groups):
+            if update_groups_for_layer:
+                # The priority queue sorts tupley lexicographically, so timestamp is the main sorting key here
+                heapq.heappush(queue, (update_groups_for_layer[0], i, 0))
+        while queue:
+            value, layer_index, element_index = heapq.heappop(queue)
+            timestamp, version, update_group = value
+            tracing_id = layers[layer_index][0]
+
+            unified_version += 1
+            version_mapping[tracing_id][version] = unified_version
+            self.save_update_group(annotation['_id'], unified_version, update_group)
+
+            if element_index + 1 < len(all_update_groups[layer_index]):
+                next_element = all_update_groups[layer_index][element_index + 1]
+                heapq.heappush(queue, (next_element, layer_index, element_index + 1))
+
+        return version_mapping
+
+    def get_editable_mapping_id(self, tracing_id: str, layer_type: str) -> Optional[str]:
+        if layer_type == "Skeleton":
+            return None
+        tracing_raw = self.get_newest_tracing_raw(tracing_id, "volumes")
+        if tracing_raw is None:
+            return None
+        volume = Volume.VolumeTracing()
+        volume.ParseFromString(tracing_raw)
+        if volume.hasEditableMapping:
+            return volume.mappingName
+        return None
+
+    def get_newest_tracing_raw(self, tracing_id, collection) -> Optional[bytes]:
+        getReply = self.src_stub.Get(
+            proto.GetRequest(collection=collection, key=tracing_id, mayBeEmpty=True)
+        )
+        if getReply.success:
+            return getReply.value
+        return None
+
+    def process_update_group(self, tracing_id: str, layer_type: str, update_group_raw: bytes, json_encoder, json_decoder) -> Tuple[bytes, int, Optional[int]]:
+        update_group_parsed = json_decoder.decode(update_group_raw)
+
+        revert_source_version = None
+        action_timestamp = 0
+
+        for update in update_group_parsed:
+            name = update["name"]
+            update_value = update["value"]
+
+            # renamings
+            if name == "updateTracing":
+                update["name"] = f"update{layer_type}Tracing"
+            elif name == "updateUserBoundingBoxes":
```

Comment:
I remember you telling me that bboxes were stored both in the skeleton and volume layers. Do these actions need to be deduplicated?

Reply:
I decided not to do this big change here to reduce the complexity of the PR. The precedence still stands, and so does the duplication. Otherwise, we would have to track all the values across all the versions and updates, and straighten everything out. I’d suggest attempting that only in combination with introducing a separate geometry layer, as we considered previously.